Add "apply on update" option to default values

m-kuhn opened the pull request:
This is an early non-functional proposal to get feedback on the general API and workflow.

The idea is to have a new checkbox on the fields properties, next to the default value which says "apply on update". If this is activated, default values will be applied again when a feature is edited, a common usage for this is a "last edited" timestamp.

On the feature form, the evaluated value will be shown directly when the feature is edited (so it can be manually adjusted, unless the field is set to read only).

What's open for discussion is what to do for geometry edits. Should this go into `QgsVectorLayer::changeGeometry()` and `QgsVectorLayer::updateFeature()` so it will be done transparently in the background? It sounds like a safe approach, but I dislike to have this react on API calls whereas the field updates would be on the form logic. Happy for inputs here.

@nyalldawson especially wondering for your feedback on the [`QString defaultValue` to `QgsDefaultValue defaultValue`](https://github.com/qgis/QGIS/pull/5218/files#diff-3225c6a3b9805892152fca25f49f0c70) change. Should be fine to do, `QgsVectorLayer::defaultValueExpression`  has been added in 3.0 so we wouldn't even have an API break there.

m-kuhn committed:
Add "apply on update" option to default values

m-kuhn committed:
Add some doxymentation

m-kuhn committed:
Add sip bindings

nyalldawson commented:
What you've got so far looks good to me!

Let me ponder on your ideas about where to do the updates...

m-kuhn committed:
Doxymentation

m-kuhn committed:
Use QgsDefaultValue throughout the code

m-kuhn commented:
> Let me ponder on your ideas about where to do the updates...

Just thought about this again, it would be preferable to do all the updates directly on api calls even for attributes.

The reason is, that the attribute form should sometimes show the current (non-updated) value for inspection. This sometimes also is desirable even when in edit mode. On the attribute table form view where features get committed automatically when changing features this gets particularly difficult.

Does that sound ok for everyone?

m-kuhn commented:
Idea on the "UX front": fields with auto-update will be read-only, the label (or a symbol next to it) yellow and a tooltip saying

> This field is updated automatically when making changes to this feature. Value preview: "[evaluated expression]".

m-kuhn committed:
Apply default values on update

m-kuhn committed:
Properly report QgsDefaultValue.__bool__ in Python

m-kuhn committed:
Fix typo in QgsFields api doc

m-kuhn committed:
Remove redundant nullptr check

m-kuhn committed:
Update cpp tests for QgsDefaultValue

m-kuhn committed:
Rename QgsField::defaultValue

It's now called QgsField::defaultValueDefinition, because it's only a
value after being evaluated

m-kuhn committed:
Add tests for QgsDefaultValue

m-kuhn committed:
Pssst... Quiet, there are people around

m-kuhn committed:
Actually apply default value

m-kuhn committed:
More default value tests

m-kuhn committed:
Default Value updates for QgsField stream operator

m-kuhn committed:
Doxygen updates for changed method names

m-kuhn committed:
Test updates for changed method name/signature

m-kuhn committed:
Add more applyOnUpdate tests

m-kuhn committed:
Fix compilation

m-kuhn committed:
Fix QgsVectorLayerUtils::createFeature

m-kuhn committed:
Fix missing import

m-kuhn committed:
Make Mr. Travis happy

m-kuhn committed:
Doxygen

m-kuhn committed:
Config interface for apply on update default values

m-kuhn committed:
Avoid recursion in apply on update default values

m-kuhn commented:
![image](https://user-images.githubusercontent.com/588407/31023600-a64942b2-a53c-11e7-8295-29f1104baad7.png)

m-kuhn commented:
Manually merged

nyalldawson commented on src/core/qgsvectorlayer.cpp:
```diff
@@ -2270,33 +2296,40 @@ bool QgsVectorLayer::changeGeometry( QgsFeatureId fid, const QgsGeometry &geom )
   {
     case QgsFields::OriginJoin:
-      return mJoinBuffer->changeAttributeValue( fid, field, newValue, oldValue );
+      result = mJoinBuffer->changeAttributeValue( fid, field, newValue, oldValue );
```
I think you're missing a "break" here (and also for the other cases modified here)

m-kuhn replied:
right, thanks!
